baybe/objectives/desirability.py

```diff
@@ -0,0 +1,135 @@
+"""Functionality for desirability objectives."""
+
+from collections.abc import Sequence
+from functools import partial
+from typing import Callable, Union
+
+import cattrs
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+from attrs import define, field
+from attrs.validators import deep_iterable, instance_of, min_len
+from typing_extensions import TypeGuard
+
+from baybe.objectives.base import Objective
+from baybe.objectives.enum import Scalarization
+from baybe.targets.base import Target
+from baybe.targets.numerical import NumericalTarget
+from baybe.utils.basic import to_tuple
+from baybe.utils.numerical import geom_mean
+
+
+def _normalize_weights(weights: Sequence[Union[float, int]]) -> tuple[float, ...]:
+    """Normalize a collection of weights such that they sum to 1.
+
+    Args:
+        weights: The un-normalized weights.
+
+    Raises:
+        ValueError: If any of the weights is non-positive.
+
+    Returns:
+        The normalized weights.
+    """
+    array = np.asarray(cattrs.structure(weights, tuple[float, ...]))
+    if not np.all(array > 0.0):
+        raise ValueError("All weights must be strictly positive.")
+    return tuple(array / array.sum())
+
+
+def _is_all_numerical_targets(
+    x: tuple[Target, ...], /
+) -> TypeGuard[tuple[NumericalTarget, ...]]:
+    """Typeguard helper function."""
+    return all(isinstance(y, NumericalTarget) for y in x)
+
+
+def scalarize(
+    values: npt.ArrayLike, scalarization: Scalarization, weights: npt.ArrayLike
+) -> np.ndarray:
+    """Scalarize the rows of a 2-D array, producing a 1-D array.
+
+    Args:
+        values: The 2-D array whose rows are to be scalarized.
+        scalarization: The scalarization mechanism to be used.
+        weights: Weights for the columns of the input array.
+
+    Raises:
+        ValueError: If the requested scalarization mechanism is not implemented.
+
+    Returns:
+        np.ndarray: A 1-D array containing the scalarized values.
+    """
+    func: Callable
+    if scalarization is Scalarization.GEOM_MEAN:
+        func = geom_mean
+    elif scalarization is Scalarization.MEAN:
+        func = partial(np.average, axis=1)
+    else:
+        raise NotImplementedError(
+            f"No scalarization mechanism defined for '{scalarization.name}'."
+        )
+    return func(values, weights=weights)
+
+
+@define(frozen=True)
+class DesirabilityObjective(Objective):
+    """An objective scalarizing multiple targets using desirability values."""
+
+    targets: tuple[Target, ...] = field(
+        converter=to_tuple,
+        validator=[min_len(2), deep_iterable(member_validator=instance_of(Target))],  # type: ignore[type-abstract]
+    )
+    "The targets considered by the objective."
+
+    weights: tuple[float, ...] = field(converter=_normalize_weights)
+    """The weights to balance the different targets.
+    By default, all targets are considered equally important."""
+
+    scalarization: Scalarization = field(
```

Comment:
@AVHopp, @Scienfitz. One thing for up for debate: I renamed the combine_func attribute to `scalarization` for the reasons already explained. Another alternative would be `scalarizer`. Let me hear your opinions 😃

Reply:
prefer `scalarizer` (also in the enum)

technically though its not correct, the scalarization is don by cscaling + combining, this just does the latter part (hence the previous name), but thats minor

Reply:
ok, so we have a tie ... do you guys want to fight it out? 😄

Reply:
you have no preference? in case of reviewer tie you're free to choose between the proposed things (I promote peace)

Reply:
I also promote peace 😆

Reply:
Ok, haven't had strong arguments for either of the two, but renamed to `scalarizer` because:
* it's shorter
* it makes sense to say "the mean is / acts as a scalarizer"